presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/DriverConnectionFactory.java

```diff
@@ -45,20 +54,41 @@ public static Properties basicConnectionProperties(BaseJdbcConfig config)
         return connectionProperties;
     }
 
-    public DriverConnectionFactory(Driver driver, String connectionUrl, Properties connectionProperties)
+    public DriverConnectionFactory(Driver driver, String connectionUrl, Optional<String> userCredentialName, Optional<String> passwordCredentialName, Properties connectionProperties)
     {
         this.driver = requireNonNull(driver, "driver is null");
         this.connectionUrl = requireNonNull(connectionUrl, "connectionUrl is null");
         this.connectionProperties = new Properties();
-        this.connectionProperties.putAll(requireNonNull(connectionProperties, "basicConnectionProperties is null"));
+        this.connectionProperties.putAll(requireNonNull(connectionProperties, "connectionProperties is null"));
+        this.userCredentialName = requireNonNull(userCredentialName, "userCredentialName is null");
+        this.passwordCredentialName = requireNonNull(passwordCredentialName, "passwordCredentialName is null");
     }
 
     @Override
     public Connection openConnection(JdbcIdentity identity)
             throws SQLException
     {
-        Connection connection = driver.connect(connectionUrl, connectionProperties);
+        Properties updatedConnectionProperties;
+        if (userCredentialName.isPresent() || passwordCredentialName.isPresent()) {
+            updatedConnectionProperties = new Properties();
+            updatedConnectionProperties.putAll(connectionProperties);
+            userCredentialName.ifPresent(credentialName -> setConnectionProperty(updatedConnectionProperties, identity.getExtraCredentials(), credentialName, "user"));
+            passwordCredentialName.ifPresent(credentialName -> setConnectionProperty(updatedConnectionProperties, identity.getExtraCredentials(), credentialName, "password"));
+        }
+        else {
+            updatedConnectionProperties = connectionProperties;
+        }
+
+        Connection connection = driver.connect(connectionUrl, updatedConnectionProperties);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *connection*. The resource is referenced at line 83. The resource is returned at line 84. There are other execution paths that don&#39;t close the resource or return it, for example, when DriverConnectionFactory.checkState throws an exception. To prevent this resource leak, close *connection* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).